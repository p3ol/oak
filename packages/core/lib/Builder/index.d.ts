import {
  AddonObject,
  ElementObject,
  BuilderOptions,
  Component,
  ComponentObject,
  ComponentsGroup,
  ComponentOverride,
  Field,
  FieldObject,
  FieldOverride,
  TextsSheet,
  TextsSheetObject,
  ComponentSettingsField,
  SettingOverride,
  ComponentOverrideObject,
  ElementId,
  ComponentSettingsFieldKeyTuple,
  GetTextCallback,
  ComponentSettingsTab,
} from '../types';
import { Logger } from '../Logger';
import { Emitter, EmitterCallback } from '../Emitter';

export declare class Builder extends Emitter {
  constructor(opts?: {
    addons?: AddonObject[];
    content?: ElementObject[];
    options?: BuilderOptions;
  });

  options: BuilderOptions;
  logger: Logger;

  subscribe(cb: EmitterCallback): (() => void);
  setAddons(addons: AddonObject[]): void;
  addAddon(addon: AddonObject): void;
  removeAddon(addon: AddonObject): void;
  getAvailableComponents(): ComponentsGroup[];
  getComponent(type: string): Component;
  getComponentDisplayableSettings(
    element: ElementObject,
    opts?: {
      component?: Component | ComponentObject;
      fields?: (Field | FieldObject)[];
    },
  ): ComponentSettingsField[];
  getAvailableFields(): Array<Field>;
  getField(type: string): Field;
  getOverride(
    type: string,
    target: string,
    opts?: {
      output?: 'field';
      setting?: ComponentSettingsField;
    },
  ): ComponentOverride | FieldOverride;
  mergeOverrides(
    overrides: (ComponentOverride | FieldOverride | SettingOverride)[]
  ): void;
  getContent(): ElementObject[];
  setContent(content: ElementObject[], opts?: { emit?: boolean }): void;
  createElement(type: string, opts: {
    component?: Component | ComponentObject;
    override?: ComponentOverride | ComponentOverrideObject;
    baseElement?: ElementObject;
    resetIds?: boolean;
  }): object;
  addElement(element: ElementObject, opts?: {
    component?: Component | ComponentObject;
    parent?: ElementObject[];
    position?: 'before' | 'after';
  }): Element;
  addElements(elements: ElementObject[], options?: {
    parent?: ElementObject[];
    position?: 'before' | 'after';
  }): ElementObject[];
  getElement(id: ElementId): ElementObject;
  removeElement(id: ElementId, opts?: {
    deep?: boolean;
  }): boolean;
  setElement(id: ElementId, updates: ElementObject, opts?: {
    deep?: boolean;
  }): ElementObject;
  moveElement(
    element: ElementObject,
    sibling: ElementObject,
    opts?: {
      deep?: boolean;
      position?: 'before' | 'after';
      parent: ElementObject;
    }
  ): ElementObject;
  duplicateElement(
    element: ElementObject,
    opts?: {
      deep?: boolean;
    }
  ): ElementObject;
  getElementSettings(
    element: ElementObject,
    key: string | string[] | ComponentSettingsFieldKeyTuple[],
    def?: any
  ): any;
  canUndo(): boolean;
  canRedo(): boolean;
  undo(): void;
  redo(): void;
  resetHistory(): void;
  generateId(): string;
  getTextSheet(id: string): TextsSheet;
  getText(key: string | GetTextCallback, def?: any): any;
  setText(key: string, value: any): void;
  getActiveTextSheet(): TextsSheet;
  setActiveTextSheet(id: string): void;
  getAvailableSettings(): (ComponentSettingsTab | ComponentSettingsField)[];
}
