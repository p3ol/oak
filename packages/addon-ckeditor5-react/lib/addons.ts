import type { ReactFieldObject, AddonObject } from '@oakjs/react';
import { omit } from '@oakjs/react';
import ClassicEditor from '@oakjs/ckeditor5-build-custom';

import Field, { CKEditorFieldProps } from './Field';

export const ckeditorField = ({
  config,
  editor,
  ...props
}: CKEditorFieldProps = {}): ReactFieldObject => ({
  type: 'ckeditor',
  render: Field,
  ...omit(props, ['onChange']),
  props: {
    editor: editor || ClassicEditor,
    config: {
      ...config,
      link: {
        decorators: {
          openInNewTab: {
            mode: 'manual',
            label: 'Open in a new tab',
            defaultValue: false,
            attributes: {
              target: '_blank',
              rel: 'noopener noreferrer',
            },
          },
        },
      },
      toolbar: {
        items: [
          'undo',
          'redo',
          '|',
          'bold',
          'italic',
          'strikethrough',
          'underline',
          'link',
          'fontColor',
          'fontSize',
          'bulletedList',
          'numberedList',
          '|',
          'removeFormat',
          '|',
          'outdent',
          'indent',
          'alignment',
          '|',
          'horizontalLine',
          'blockQuote',
          'insertTable',
        ],
      },
    },
  },
});

export const ckeditorFieldAddon = (
  props?: CKEditorFieldProps
): AddonObject => ({
  fields: [ckeditorField(props)],
});
