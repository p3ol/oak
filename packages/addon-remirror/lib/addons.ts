import type { AddonObject, ReactFieldObject } from '@oakjs/react';
import {
  BoldExtension,
  FontSizeExtension,
  ItalicExtension,
  UnderlineExtension,
  LinkExtension,
  TextColorExtension,
  NodeFormattingExtension,
} from 'remirror/extensions';

import type { Extensions } from './types';
import Field, { type RemirrorFieldProps } from './Field';

export const remirrorField = ({
  extensions,
  ...props
}: RemirrorFieldProps = {}): ReactFieldObject => ({
  type: 'remirror',
  render: Field,
  ...props,
  props: {
    extensions,
  },
});

export const basicExtensions = (): Extensions[] => [
  new BoldExtension({}),
  new ItalicExtension(),
  new UnderlineExtension(),
  new LinkExtension({}),
  new TextColorExtension({}),
  new FontSizeExtension({
    defaultSize: '16px',
    unit: 'px',
  }),
  new NodeFormattingExtension({}),
];

export const remirrorFieldAddon = (props?: ReactFieldObject): AddonObject => ({
  fields: [remirrorField({
    extensions: basicExtensions,
    ...props,
  })],
});
