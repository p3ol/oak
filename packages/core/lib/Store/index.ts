import { exists, get, set, cloneDeep } from '@junipero/core';

import type {
  ComponentObject,
  ComponentOverrideObject,
  ComponentSettingsFieldKeyTuple,
  ElementId,
  ElementObject,
  ElementSettingsComplexKey,
  ElementSettingsKeyObject,
  StoreSanitizeOptions,
  StoreFindOptions,
  StoreFindDeepOptions,
} from '../types';
import {
  Component,
  ComponentOverride,
} from '../classes';
import Emitter from '../Emitter';
import Builder from '../Builder';

export declare abstract class IStore {
  constructor(options?: { builder: Builder });

  /**
   * Check if an element id is valid.
   * An Element ID is a string or a number, and cannot be empty.
   */
  isIdValid(id: string | number): boolean;

  /** Check if two element ids are the same */
  isSameElement(elementId: ElementId, siblingId: ElementId): boolean;

  /** Get the content of the store */
  get(): Array<ElementObject>;

  /**
   * Set the content of the store.
   * If the emit option is set to false, the store will not emit a change event.
   */
  set(content: Array<ElementObject>, options?: { emit?: boolean }): void;

  /**
   * Sanitize an element object (adds missing properties, ids, etc.)
   * If the resetIds option is set to true, the element id will be reset even
   * if it already exists.
   */
  sanitize(
    element: ElementObject,
    options?: StoreSanitizeOptions
  ): ElementObject;

  /** Creates a new element object based on an existing component (or not) */
  createElement(type: string, options?: Partial<{
    baseElement?: ElementObject;
    component?: ComponentObject;
    override?: ComponentOverrideObject;
  } & StoreSanitizeOptions>): ElementObject;

  /** Adds an element to the store */
  addElement(element: ElementObject, options?: Partial<{
    position?: 'before' | 'after';
  } & StoreFindOptions & StoreSanitizeOptions>): ElementObject;

  /** Adds multiple elements to the store */
  addElements(elements: Array<ElementObject>, options?: Partial<{
    position?: 'before' | 'after';
  } & StoreFindOptions & StoreSanitizeOptions>): Array<ElementObject>;

  /** Finds an element in the store */
  getElement(id: ElementId, options?: StoreFindDeepOptions): ElementObject;

  /** Removes an element from the the store */
  removeElement(id: ElementId, options?: StoreFindDeepOptions): boolean;

  /** Updates an element in the store with new props */
  setElement(
    id: ElementId,
    newContent: object,
    options?: StoreFindDeepOptions
  ): ElementObject;

  /** Moves an element next to a sibling (inside the same parent or not) */
  moveElement(
    element: ElementObject,
    sibling: ElementObject,
    options?: Partial<StoreFindOptions & {
      position?: 'before' | 'after';
    }>
  ): ElementObject;

  /** Duplicate an element */
  duplicateElement(
    element: ElementObject,
    options?: StoreFindOptions
  ): ElementObject;

  /** Recursively finds the nearest parent of an element */
  findNearestParent(
    id: ElementId,
    options?: StoreFindOptions
  ): Array<ElementObject>;

  /** Recursively checks if an element is inside a parent */
  contains(id: ElementId, options?: StoreFindOptions): boolean;

  /** Retrieves the setting value of an element */
  getElementSettings(
    element: ElementObject,
    key: string | Array<string> | Array<ComponentSettingsFieldKeyTuple>,
    def?: any
  ): any;

  /** Sets the setting value of an element */
  setElementSettings(
    element: ElementObject,
    key: string | Array<string> | Array<ComponentSettingsFieldKeyTuple>,
    value: any
  ): void;

  /** Commit changes into history */
  commit(): void;

  /** Undo the last change */
  undo(): void;

  /** Redo the last change */
  redo(): void;

  canUndo(): boolean;
  canRedo(): boolean;
  resetHistory(): void;
}

export default class Store extends Emitter implements IStore {
  #content: ElementObject[] = [];
  #history: ElementObject[] = [];
  #historyIndex: number = 0;
  #builder: Builder = null;

  constructor ({ builder }: { builder: Builder }) {
    super();

    this.#builder = builder;
  }

  isIdValid (id: ElementId) {
    return exists(id) && id !== '';
  }

  isSameElement (elementId: ElementId, siblingId: ElementId) {
    return this.isIdValid(elementId) && this.isIdValid(siblingId) &&
      elementId === siblingId;
  }

  get () {
    return this.#content;
  }

  set (content: Array<ElementObject>, { emit = true } = {}) {
    this.#content = (Array.isArray(content) ? content : [])
      .map(e => this.sanitize(e, { withDefaults: true }));

    emit && this.emit('content.update', this.#content);
    this.commit();
  }

  sanitize (element: ElementObject, {
    component: c,
    override: o,
    ...opts
  }: {
    component?: Component,
    override?: ComponentOverride,
    [_: string]: any
  } = {}) {
    if (!this.isIdValid(element.id) || opts.resetIds) {
      element.id = this.#builder.generateId();
    }

    const component = c || this.#builder.getComponent(element.type);
    const override = o || this.#builder.getOverride(
      'component', element.type
    ) as ComponentOverrideObject;
    const override_ = new ComponentOverride(
      override as ComponentOverrideObject
    );

    if (opts.withDefaults) {
      element = {
        ...(override_?.construct || component?.construct)?.({
          builder: this.#builder,
        }) || {},
        ...element,
      };
    }

    const deserialize = override_?.deserialize || component?.deserialize;
    element = deserialize?.(element, { builder: this.#builder }) || element;

    const customSanitize = override_?.sanitize || component?.sanitize;
    element = customSanitize?.(element, { builder: this.#builder }) || element;

    const containers = override_?.getContainers?.(element) ||
      component?.getContainers?.(element) ||
      [element.content];

    containers.forEach((container: ElementObject[]) => {
      if (Array.isArray(container)) {
        container.forEach((elmt, i) => {
          container[i] = this.sanitize(elmt, opts);
        });
      }
    });

    return element;
  }

  createElement (type: string, {
    component: c,
    override: o,
    baseElement,
    ...opts
  }: {
    component?: ComponentObject,
    override?: ComponentOverrideObject,
    baseElement?: ElementObject,
    [_: string]: any
  } = {}) {
    const component = c
      ? new Component(c) : new Component(this.#builder.getComponent(type));
    const override = o
      ? new ComponentOverride(o as ComponentOverrideObject)
      : this.#builder.getOverride(
        'component',
        component.id
      ) as ComponentOverride;
    baseElement = {
      ...baseElement,
      ...component.construct?.({
        builder: this.#builder,
      }) || {},
    };
    const elementWithContent = {
      ...baseElement,
      content: typeof baseElement.content === 'function'
        ? baseElement.content(this.#builder.getText.bind(this.#builder))
        : baseElement.content,
    };

    const element = {
      ...elementWithContent,
      ...(override?.construct?.({
        builder: this.#builder,
        baseElement: elementWithContent,
      }) || {}),
    };

    return this.sanitize(element, { component, override, ...opts });
  }

  addElement (element: ElementObject, {
    parent = this.#content,
    position = 'after',
    component,
    ...opts
  }: {
    parent?: Array<ElementObject>,
    position?: 'before' | 'after',
    component?: ComponentObject,
    [_: string]: any
  } = {}) {
    this.#builder.logger.log('Adding element:', element, { parent, position });

    // If component is provided, we construct the element from it
    // Otherwise, we just sanitize it
    element = component
      ? this.createElement(component.id, { component, ...opts })
      : this.sanitize(element, opts);

    switch (position) {
      case 'before':
        parent.unshift(element);
        break;
      case 'after':
        parent.push(element);
        break;
    }

    this.commit();
    this.emit('content.update', this.#content);

    return element;
  }

  addElements (elements: Array<ElementObject>, {
    parent = this.#content,
    position = 'after',
    ...opts
  } = {}) {
    this.#builder.logger
      .log('Adding elements:', elements, { parent, position });

    elements = elements.map(e => this.sanitize(e, opts));

    switch (position) {
      case 'before':
        parent.unshift(...elements);
        break;
      case 'after':
        parent.push(...elements);
        break;
    }

    this.commit();
    this.emit('content.update', this.#content);

    return elements;
  }

  getElement (
    id: string,
    { parent = this.#content, deep = false }: {
      deep?: boolean,
      parent?: ElementObject[]
    } = {}
  ): ElementObject {
    if (!this.isIdValid(id)) {
      return;
    }

    const found = parent.find(elmt => this.isSameElement(elmt?.id, id));

    if (found) {
      return found;
    }

    if (deep) {
      for (const child of parent) {
        const component = this.#builder.getComponent(child.type);
        const override = this.#builder.getOverride('component', child.type);
        const containers: ElementObject[][] =
          (override as ComponentOverride)?.getContainers?.(child) ||
          component?.getContainers?.(child) ||
          [child.content as ElementObject[]];

        for (const container of containers) {
          const found = this.getElement(id, { parent: container, deep });

          if (found) {
            return found;
          }
        }
      }
    }
  }

  removeElement (
    id: string, {
      parent = this.#content,
      deep,
    }: {
      parent?: Array<ElementObject>,
      deep?: boolean
    } = {}) {
    if (!this.isIdValid(id)) {
      return;
    }

    const index = parent.findIndex(elmt => this.isSameElement(elmt?.id, id));

    if (index > -1) {
      parent.splice(index, 1);
      this.commit();
      this.emit('content.update', this.#content);

      return true;
    }

    if (deep) {
      for (const child of parent) {
        const component = this.#builder.getComponent(child.type);
        const override = this.#builder.getOverride('component', child.type);
        const containers =
          (override as ComponentOverride)?.getContainers?.(child) ||
          component?.getContainers?.(child) ||
          [child.content as ElementObject[]];

        for (const container of containers) {
          const removed = this.removeElement(id, { parent: container, deep });

          if (removed) {
            return true;
          }
        }
      }
    }

    return false;
  }

  setElement (id: string, newContent: Partial<ElementObject>, {
    element: e,
    parent = this.#content,
    deep,
  }: {
    element?: ElementObject,
    parent?: Array<ElementObject>,
    deep?: boolean
  } = {}) {
    if (!this.isIdValid(id)) {
      return;
    }

    const element = e || this.getElement(id, { parent, deep });
    const component = this.#builder.getComponent(element.type);
    const override: ComponentOverride =
      this.#builder.getOverride('component', element.type) as ComponentOverride;

    const serialize = override?.serialize || component?.serialize;

    Object.assign(
      element,
      serialize?.(newContent, { builder: this.#builder }) ||
        newContent || {}
    );

    this.commit();
    this.emit('content.update', this.#content);

    return element;
  }

  moveElement (
    element?: ElementObject,
    sibling?: ElementObject,
    {
      parent = this.#content,
      position,
    }: {
      parent?: Array<ElementObject>,
      position?: 'before' | 'after'
    } = {}) {
    if (
      this.isSameElement(element?.id, sibling?.id) ||
      this.contains(sibling.id, { parent: element })
    ) {
      return;
    }

    const nearestParent = this.findNearestParent(element.id);
    const childIndex = nearestParent
      ?.findIndex(e => this.isSameElement(e?.id, element.id));
    const [retrievedElement] = nearestParent?.splice(childIndex, 1) || [];

    if (!retrievedElement) {
      return;
    }

    const newChildIndex = parent.indexOf(sibling);
    parent.splice(
      position === 'after' ? newChildIndex + 1 : newChildIndex,
      0,
      retrievedElement
    );

    this.commit();
    this.emit('content.update', this.#content);

    return retrievedElement;
  }

  duplicateElement (element: ElementObject, { parent = this.#content } = {}) {
    let newElmt = this.sanitize(cloneDeep(element), { resetIds: true });
    const component = this.#builder.getComponent(element.type);
    const override: ComponentOverride =
      this.#builder.getOverride('component', element.type) as ComponentOverride;
    const duplicate = override?.duplicate || component?.duplicate;

    if (typeof duplicate === 'function') {
      newElmt = duplicate(newElmt);
    }

    parent.splice(
      parent.findIndex(e => this.isSameElement(e.id, element.id)) + 1,
      0,
      newElmt
    );

    this.commit();
    this.emit('content.update', this.#content);

    return newElmt;
  }

  findNearestParent (
    id: string | ElementId, { parent = this.#content } = {}
  ): Array<ElementObject> {
    // First check if element in inside direct parent to avoid trying to
    // find every component & override for every nested level
    for (const e of parent) {
      if (this.isSameElement(e.id, id)) {
        return parent;
      }
    }

    // Then try to find the element inside nested containers
    // Row component for exemple doesn't have a content property and uses
    // the getContainers component method to return the nested columns
    // (using the cols property)
    for (const e of parent) {
      const component = this.#builder.getComponent(e.type);
      const override: ComponentOverride =
        this.#builder.getOverride('component', e.type) as ComponentOverride;

      const containers = override?.getContainers?.(e) ||
        component?.getContainers?.(e) ||
        [e.content];

      for (const container of containers) {
        if (Array.isArray(container)) {
          const nearest = this.findNearestParent(id, {
            parent: container,
          });

          if (nearest) {
            return nearest;
          }
        }
      }
    }

    return null;
  }

  contains (
    id: string | ElementId,
    { parent = this.#content }: { parent?: ElementObject | Array<any> } = {}
  ) {
    // Force parent to be an array to be able to loop over it
    // -----
    // In some cases (Store.moveElement for example), parent cannot be
    // `element.content` because it would prevent rows (with `element.cols`)
    // or other container-based elements from being checked by `contains`
    const parent_ = [].concat(parent);

    for (const e of parent_) {
      if (this.isSameElement(e?.id, id)) {
        return true;
      }
    }

    for (const e of parent_) {
      const component = this.#builder.getComponent(e.type);
      const override: ComponentOverride =
        this.#builder.getOverride('component', e.type) as ComponentOverride;

      const containers = override?.getContainers?.(e) ||
        component?.getContainers?.(e) ||
        [e.content];

      for (const container of containers) {
        if (Array.isArray(container)) {
          if (this.contains(id, { parent: container })) {
            return true;
          }
        }
      }
    }

    return false;
  }

  // This allows 4 signatures:
  // (element, 'settings.size')
  // (element, { from: 'size', to: 'settings.size' })
  // (element, [{ from: 'size', to: 'settings.size' }])
  // (element, ['settings.size', 'settings.url'])
  getElementSettings (
    element: ElementObject,
    key: ElementSettingsKeyObject,
    def?: string
  ) {
    if (Array.isArray(key)) {
      return key.reduce((
        res: any,
        k: { from: string, to: string, default: string} | string
      ) => {
        res[(k as ElementSettingsComplexKey)?.to || k as string] =
          get(
            element,
            (k as ElementSettingsComplexKey)?.from || k as string,
            (k as ElementSettingsComplexKey)?.default ?? def as string
          );

        return res;
      }, {});
    } else {
      return get(
        element,
        (key as ElementSettingsComplexKey)?.from || key as string,
        (key as ElementSettingsComplexKey)?.default ?? def
      );
    }
  }

  // This allows 4 signatures:
  // (element, 'settings.size', 20)
  // (element, { from: 'size', to: 'settings.size' }, { size: 20 })
  // (element, [{ from: 'size', to: 'settings.size' }], { size: 20 })
  // (element, ['size', 'url'], { size: 20, url: '...' })
  setElementSettings (
    element: ElementObject,
    key: ElementSettingsKeyObject | string,
    value: any
  ) {
    if (Array.isArray(key)) {
      key.forEach(k =>
        set(
          element,
          (k as ElementSettingsComplexKey)?.to || k as string,
          get(value, (k as ElementSettingsComplexKey)?.from || k as string,
            (k as ElementSettingsComplexKey).default)
        )
      );
    } else {
      set(
        element,
        (key as ElementSettingsComplexKey)?.to || key as string,
        (key as ElementSettingsComplexKey)?.from ? get(
          value,
          (key as ElementSettingsComplexKey).from,
          (key as ElementSettingsComplexKey)?.default
        ) : value ?? (key as ElementSettingsComplexKey)?.default
      );
    }

    return element;
  }

  commit () {
    if (this.#historyIndex > 0) {
      this.#history = this.#history.slice(this.#historyIndex);
    }

    this.#historyIndex = 0;
    this.#history.unshift(cloneDeep(this.#content));
    this.#history = this.#history.slice(0, this.#builder.options.historyLimit);
    this.emit('history.commit', this.#history, this.#historyIndex);
  }

  canUndo () {
    return this.#historyIndex < this.#history.length - 1;
  }

  canRedo () {
    return this.#historyIndex > 0;
  }

  undo () {
    if (!this.canUndo()) {
      return;
    }

    this.#historyIndex++;
    this.#content = cloneDeep(this.#history[this.#historyIndex]);
    this.emit('history.undo', this.#history, this.#historyIndex);
    this.emit('content.update', this.#content);
  }

  redo () {
    if (!this.canRedo()) {
      return;
    }

    this.#historyIndex--;
    this.#content = cloneDeep(this.#history[this.#historyIndex]);
    this.emit('history.redo', this.#history, this.#historyIndex);
    this.emit('content.update', this.#content);
  }

  resetHistory () {
    this.#history = [cloneDeep(this.#content)];
    this.#historyIndex = 0;
    this.emit('history.reset', this.#history, this.#historyIndex);
  }
}
