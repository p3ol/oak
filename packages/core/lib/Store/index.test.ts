import { ComponentOverride } from '../classes';
import { baseAddon, titleComponent } from '../addons';
import Builder from '../Builder';
import Store from './index';

describe('Store', () => {
  const getBuilder = (opts?: ConstructorParameters<typeof Builder>[0]) => {
    let id = 0;

    return new Builder({
      ...opts,
      options: {
        ...opts?.options,
        generateId: () => id++,
      },
    });
  };

  it('should allow to manage a set of elements', () => {
    const content = [
      { type: 'title', content: 'This is a title' },
      { type: 'text', content: 'This is a text' },
    ];

    const store = new Store({ builder: getBuilder() });
    store.set(content);

    expect(store.get()).toMatchSnapshot('Base content');

    store.removeElement(1);

    expect(store.get()).toMatchSnapshot('Without text');

    const button = { type: 'button', content: 'This is a button' };
    store.addElement(button);

    expect(store.get()).toMatchSnapshot('With button');

    store.setElement(2, { content: 'This is a new button' });

    expect(store.get()).toMatchSnapshot('With button updated');

    store.moveElement(button, content[0], { position: 'before' });

    expect(store.get()).toMatchSnapshot('With button & title swapped');

    store.duplicateElement(button);

    expect(store.get()).toMatchSnapshot('With button duplicated');
  });

  describe('.sanitize()', () => {
    it('should allow to sanitize an element', () => {
      const store = new Store({ builder: getBuilder() });
      expect(store.sanitize({ type: 'title', content: 'This is a title' }))
        .toMatchSnapshot('Add id');

      expect(store.sanitize(
        { type: 'title' },
        { withDefaults: true, component: titleComponent() }
      )).toMatchSnapshot('Add id & default props from component');

      expect(store.sanitize(
        { type: 'title', content: 'This is a title' },
        {
          withDefaults: true,
          component: titleComponent({
            sanitize: element => ({
              ...element,
              content: element.content.toString() + ', a not so great one',
            }),
          }),
          override: new ComponentOverride({
            sanitize: element => ({
              ...element,
              content: element.content.toString() + ', a really great one',
            }),
          }),
        }
      )).toMatchSnapshot(
        'Allow custom sanitize on component or override level'
      );
    });

    it('should allow to sanitize elements with custom containers', () => {
      const store = new Store({ builder: getBuilder({
        addons: [baseAddon()],
      }) });

      expect(store.sanitize(
        { type: 'row', cols: [
          { type: 'col', content: [
            { type: 'title' },
          ] },
        ] },
        { withDefaults: true }
      )).toMatchSnapshot('Deeply sanitize');
    });
  });

  describe('.createElement()', () => {
    it('should allow to create an element', () => {
      const store = new Store({ builder: getBuilder({
        addons: [baseAddon()],
      }) });

      expect(store.createElement('title')).toMatchSnapshot('Title');
      expect(store.createElement('title', {
        override: new ComponentOverride({
          construct: () => ({ content: 'This is a great title' }),
        }),
      })).toMatchSnapshot('Title with custom construct');
    });
  });

  describe('.getElement()', () => {
    it('should allow to get an element', () => {
      const store = new Store({ builder: getBuilder() });
      store.set([{ type: 'title', content: 'This is a title' }]);

      expect(store.getElement(0)).toMatchSnapshot('Shallow element');
    });

    it('should allow to get an element with custom containers', () => {
      const store = new Store({ builder: getBuilder({
        addons: [baseAddon()],
      }) });
      store.set([{ type: 'row', cols: [
        { type: 'col', content: [
          { id: 'my-title', type: 'title', content: 'This is another title' },
        ] },
      ] }]);

      expect(store.getElement('my-title', { deep: true }))
        .toMatchSnapshot('Deep element');
    });
  });

  describe('.findNearestParent()', () => {
    it('should allow to find the nearest parent of an element', () => {
      const store = new Store({ builder: getBuilder({
        addons: [baseAddon()],
      }) });
      store.set([{ id: 'my-row', type: 'row', cols: [
        { id: 'my-col', type: 'col', content: [
          { id: 'my-title', type: 'title', content: 'This is another title' },
        ] },
      ] }]);

      expect(store.findNearestParent('my-row')).toMatchSnapshot('content');
      expect(store.findNearestParent('my-col')).toMatchSnapshot('row.cols');
      expect(store.findNearestParent('my-title'))
        .toMatchSnapshot('col.content');
    });
  });

  describe('.contains()', () => {
    it('should allow to check if an element is contained in the store', () => {
      const store = new Store({ builder: getBuilder({
        addons: [baseAddon()],
      }) });
      store.set([{ id: 'my-row', type: 'row', cols: [
        { id: 'my-col', type: 'col', content: [
          { id: 'my-title', type: 'title', content: 'This is another title' },
        ] },
      ] }]);

      expect(store.contains('my-row')).toBe(true);
      expect(store.contains('my-col')).toBe(true);
      expect(store.contains('my-title')).toBe(true);
      expect(store.contains('my-button')).toBe(false);
    });
  });

  describe('.getElementSettings()', () => {
    it('should allow to get the settings of an element', () => {
      const store = new Store({ builder: getBuilder({
        addons: [baseAddon()],
      }) });
      const element = {
        type: 'image',
        url: 'https://example.com/image.jpg',
        name: 'My image',
        settings: {
          width: 100,
          height: 100,
          rotate: 0,
        },
      };

      expect(store.getElementSettings(element, 'url')).toBe(element.url);
      expect(store.getElementSettings(element, 'settings.width')).toBe(100);
      expect(store.getElementSettings(element, 'settings.rotate')).toBe(0);
      expect(store
        .getElementSettings(element, [{ from: 'settings.width', to: 'width' }])
      ).toMatchObject({ width: 100 });
    });
  });

  describe('.setElementSettings()', () => {
    it('should allow to set the settings of an element', () => {
      const store = new Store({ builder: getBuilder({
        addons: [baseAddon()],
      }) });
      const element = {
        type: 'image',
        url: 'https://example.com/image.jpg',
        name: 'My image',
        settings: {
          width: 100,
          height: 100,
          rotate: 0,
        },
      };

      store
        .setElementSettings(element, 'url', 'https://example.com/image2.jpg');
      expect(element.url).toBe('https://example.com/image2.jpg');

      store.setElementSettings(element, 'settings.width', 200);
      expect(element.settings.width).toBe(200);

      store.setElementSettings(element, 'settings.rotate', 90);
      expect(element.settings.rotate).toBe(90);

      store.setElementSettings(element, [
        'url',
        { from: 'width', to: 'settings.width' },
      ], { url: 'https://example.com/image3.jpg', width: 300 });
      expect(element.url).toBe('https://example.com/image3.jpg');
      expect(element.settings.width).toBe(300);
    });
  });
});
