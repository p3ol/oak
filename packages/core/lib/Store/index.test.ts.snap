// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store .createElement() should allow to create an element: Title 1`] = `
{
  "content": "This is a title",
  "headingLevel": "h1",
  "id": 0,
  "type": "title",
}
`;

exports[`Store .createElement() should allow to create an element: Title with custom construct 1`] = `
{
  "content": "This is a great title",
  "headingLevel": "h1",
  "id": 1,
  "type": "title",
}
`;

exports[`Store .findNearestParent() should allow to find the nearest parent of an element: col.content 1`] = `
[
  {
    "content": "This is another title",
    "headingLevel": "h1",
    "id": "my-title",
    "type": "title",
  },
]
`;

exports[`Store .findNearestParent() should allow to find the nearest parent of an element: content 1`] = `
[
  {
    "cols": [
      {
        "content": [
          {
            "content": "This is another title",
            "headingLevel": "h1",
            "id": "my-title",
            "type": "title",
          },
        ],
        "id": "my-col",
        "style": {},
        "type": "col",
      },
    ],
    "id": "my-row",
    "settings": {
      "alignItems": "flex-start",
    },
    "type": "row",
  },
]
`;

exports[`Store .findNearestParent() should allow to find the nearest parent of an element: row.cols 1`] = `
[
  {
    "content": [
      {
        "content": "This is another title",
        "headingLevel": "h1",
        "id": "my-title",
        "type": "title",
      },
    ],
    "id": "my-col",
    "style": {},
    "type": "col",
  },
]
`;

exports[`Store .getElement() should allow to get an element with custom containers: Deep element 1`] = `
{
  "content": "This is another title",
  "headingLevel": "h1",
  "id": "my-title",
  "type": "title",
}
`;

exports[`Store .getElement() should allow to get an element: Shallow element 1`] = `
{
  "content": "This is a title",
  "id": 0,
  "type": "title",
}
`;

exports[`Store .sanitize() should allow to sanitize an element: Add id & default props from component 1`] = `
{
  "content": "This is a title",
  "headingLevel": "h1",
  "id": 1,
  "type": "title",
}
`;

exports[`Store .sanitize() should allow to sanitize an element: Add id 1`] = `
{
  "content": "This is a title",
  "id": 0,
  "type": "title",
}
`;

exports[`Store .sanitize() should allow to sanitize an element: Allow custom sanitize on component or override level 1`] = `
{
  "content": "This is a title, a really great one",
  "headingLevel": "h1",
  "id": 2,
  "type": "title",
}
`;

exports[`Store .sanitize() should allow to sanitize elements with custom containers: Deeply sanitize 1`] = `
{
  "cols": [
    {
      "content": [
        {
          "content": "This is a title",
          "headingLevel": "h1",
          "id": 4,
          "type": "title",
        },
      ],
      "id": 2,
      "style": {},
      "type": "col",
    },
  ],
  "id": 0,
  "settings": {
    "alignItems": "flex-start",
  },
  "type": "row",
}
`;

exports[`Store should allow to manage a set of elements: Base content 1`] = `
[
  {
    "content": "This is a title",
    "id": 0,
    "type": "title",
  },
  {
    "content": "This is a text",
    "id": 1,
    "type": "text",
  },
]
`;

exports[`Store should allow to manage a set of elements: With button & title swapped 1`] = `
[
  {
    "content": "This is a new button",
    "id": 2,
    "type": "button",
  },
  {
    "content": "This is a title",
    "id": 0,
    "type": "title",
  },
]
`;

exports[`Store should allow to manage a set of elements: With button 1`] = `
[
  {
    "content": "This is a title",
    "id": 0,
    "type": "title",
  },
  {
    "content": "This is a button",
    "id": 2,
    "type": "button",
  },
]
`;

exports[`Store should allow to manage a set of elements: With button duplicated 1`] = `
[
  {
    "content": "This is a new button",
    "id": 2,
    "type": "button",
  },
  {
    "content": "This is a new button",
    "id": 3,
    "type": "button",
  },
  {
    "content": "This is a title",
    "id": 0,
    "type": "title",
  },
]
`;

exports[`Store should allow to manage a set of elements: With button updated 1`] = `
[
  {
    "content": "This is a title",
    "id": 0,
    "type": "title",
  },
  {
    "content": "This is a new button",
    "id": 2,
    "type": "button",
  },
]
`;

exports[`Store should allow to manage a set of elements: Without text 1`] = `
[
  {
    "content": "This is a title",
    "id": 0,
    "type": "title",
  },
]
`;
